extmod/uasyncio/__init__.py

```diff
@@ -0,0 +1,9 @@
+from .core import *
+try:
+    from .event import *
+except:
+    pass
+try:
+    from .lock import *
```

Comment:
Separation into different modules is nice, but ultimately not very useful in terms of RAM usage if every module gets imported anyway. I'd suggest seperating the primitives without importing them by default. 
If a user needs those, he can just import it at the start of his project and the primitives will automatically hook into uasyncio so they can be used as expected. 
That's the approach in Peter Hinch's repo with the uasyncio suggestions: https://github.com/peterhinch/micropython-samples/blob/d8a1810a9095fa0a2b3c3e875440a15191601659/uasyncio_iostream/uasyncio/lock.py#L55

Reply:
> Separation into different modules is nice, but ultimately not very useful in terms of RAM usage if every module gets imported anyway.

The idea is that you wouldn't include the source code (either frozen or on the filesystem) for the components you don't want/need, and then these imports will just do nothing.  In other words, everything included as frozen code or on the filesystem becomes part of the library.  Doing it this way means it's compatible with CPython, eg `import uasyncio` then `uasyncio.Lock()`.

Reply:
Personally I use Locks and Events everywhere so this wouldn't impact me anyway but generally, I don't think it's a good point to expect people to remove a source file to safe some RAM. The Cpython incompatibility by importing the Lock and Event explicitly when needed is in my opinion an acceptable one. Also one that is very quickly apparent since the first test of a script will already reveal the problem.

Reply:
This isn't ideal if you are testing multiple projects on the same piece of hardware. A project not using the primitives would suffer the RAM penalty needlessly - unless you fiddled with the filesystem every time you started the code.

This is a case where the small loss of CPython compatibility offers a practical benefit.

Reply:
How big is the RAM penalty?

Reply:
On platforms like ESP8266 RAM is severely limited.

CPython provides `Lock`, `Event`, `Semaphore`, `BoundedSemaphore` , `Queue` and `Condition`. The code in this PR offers only the first two. I have suggested code for the rest and hope we will get some implementation of all of them: an aim is CPython compatibility.

In the absence of code in the PR it clearly isn't possible to do a realistic measurement of the overhead. But this is supposed to be a "Micro" implementation: for the sake of an `import` statement it seems wasteful to pull up to six unused primitives into RAM.

Reply:
Sorry for the noise, GitHub seems to not support mail reply to code comments.

How about we delegate that choice  between compatibility and saving RAM to users? We could making the RAM saving be the default but allow a build option to choose compatibility. I think I could easily add a sed script to the makefile that, depending on env vars, removes one level of `#` comment from each line between two marker lines like

```python
##begin_optional:5332:compat##
# try:
#   # …
##endof_optional:5332:compat##
```

Reply:
Some more points to think about:
A long term goal is to make uasyncio available to driver implementations in C (as far as I remember).  If one of those drivers needs to use a uasyncio.Lock, it will fail if the user removed that source file because he wants to save some RAM and doesn't use Locks in his project.

That means this approach makes it very difficult for the user/developer to figure out if he can safely remove the lock source file or not. And if you got a dynamic project (like I do) where different configurations load different modules, you can't remove any uasyncio source files because one of those project modules will surely need it. But most configurations might not need 90% of the optional uasyncio modules. So for 90% of the project configurations I might be wasting RAM on uasyncio modules that are not being used. This gets even worse the more uasyncio modules you add.

It makes a lot more sense to explicitly import the uasyncio module you need, instead of having to manually remove the source file if you don't need it.
Issuing a single "from uasyncio import lock" to load the lock module is a small price to pay for a big improvement in RAM usage and convenience.

Reply:
> That means this approach makes it very difficult for the user/developer to figure out if he can safely remove the lock source file or not. And if you got a dynamic project (like I do) where different configurations load different modules, you can't remove any uasyncio source files because one of those project modules will surely need it.

That's a good point, the possibility of a dynamic project that may or may not use some components based on a run-time decision.

> Issuing a single "from uasyncio import lock" to load the lock module is a small price to pay for a big improvement in RAM usage and convenience.

I still feel strongly about having CPython compatibility for basic usage.  It'd be better if the default (ie `import uasyncio` was to import everything, but doing something special could turn off certain features (eg `from uasyncio import minimal_core`).  But there's no way to do the latter because if there is a `__init__.py` file then it's always imported first and this needs to import all sub-components to get the former behaviour.

I have some other proposals which I'll mention below.

Reply:
I like the lazy submodule import. It solves all the problems mentioned.
Only downside is that pycharm (and other IDEs probably) aren't able to parse the lazy imported submodules, so you don't get any code completion or suggestions for those. This however is just a minor inconvenience.

Reply:
> Only downside is that pycharm (and other IDEs probably) aren't able to parse the lazy imported submodules, so you don't get any code completion or suggestions for those. This however is just a minor inconvenience.

I think this lock of code-completion is a (minor) compromise that must be made.  If lazy imports are used often (even in CPython) then the tools/IDEs could be improved to understand lazy imports.